### PR TITLE
Adding type annotations & mypy call to check them

### PR DESCRIPTION
**Pull request by happz**
Basic enhancement - it passes mypy tests without too much whitelisting
or using Any. Next steps will try to deal with Any and similar "way too
open" stuff, but otherwise, after this patch it's possible to use mypy
against projects using gluetool.

### CODE REVIEW

**Comment by happz** [comment]
Ooops, I failed to run tests :) I knew they are going to fail massively but failed to fix them.

File: gluetool/__init__.pyi [hunk 1]
```diff
@@ -0,0 +1,3 @@
+from .glue import Failure as Failure, Glue as Glue, GlueCommandError as GlueCommandError, GlueError as GlueError, GlueRetryError as GlueRetryError, Module as Module
```
Comment (mkluson) [inline gluetool/__init__.pyi]: I am not sure why do you use `import as` when you do not rename the import

File: gluetool/glue.py [hunk 2]
```diff
@@ -319,23 +373,32 @@ class Configurable(object):
...
 
-    options_note = None
+    options_note = None  # type: str
     """If set, it will be printed after all options as a help's epilog."""
 
     supported_dryrun_level = DryRunLevels.DEFAULT
     """Highest supported level of dry-run."""
 
-    unique_name = None
+    name = None  # type: str
```
Comment (mkluson) [inline gluetool/glue.py:390]: I thing variables which can be `None` should be Optimal[]

File: gluetool/glue.py [hunk 3]
```diff
@@ -319,23 +373,32 @@ class Configurable(object):
...
+    """
+    Unque name of this (module) instance.
+
+    Used by modules, has no meaning elsewhere, but since dry-run checks are done on this level,
+    it must be declared here to make pylint happy :/
     """
 
     @staticmethod
     def _for_each_option(callback, options):
+        # type: (Callable[..., None], Any) -> None
```
Comment (mkluson) [inline gluetool/glue.py:405]: What does the dots exactly means?

File: gluetool/glue.py [hunk 4]
```diff
@@ -583,7 +684,21 @@ def check_required_options(self):
             if name not in self._config or not self._config[name]:
                 raise GlueError("Missing required '{}' option".format(name))
 
+    # pylint: disable=function-redefined
```
Comment (mkluson) [inline gluetool/glue.py:701]: Why? please add comment

File: gluetool/glue.py [hunk 5]
```diff
@@ -664,11 +791,13 @@ def check_dryrun(self):
             return
 
         if self.dryrun_level > self.supported_dryrun_level:
-            raise GlueError("Module '{}' does not support current dry-run level of '{}'".format(self.unique_name,
-                                                                                                self.dryrun_level.name))
+            dryrun_level_name = self.dryrun_level.name  # type: ignore  # `dryrun_level` is not pure int but Enum
```
Comment (mkluson) [inline gluetool/glue.py:808]: Why you cannot use # type: Enum?

File: gluetool/log.py [hunk 6]
```diff
@@ -411,13 +477,24 @@ class ContextAdapter(logging.LoggerAdapter):
         the information about context.
     """
 
+    #verbose = None  # type: LoggingFunctionType
```
Comment (mkluson) [inline gluetool/log.py]: this looks like you forgot to delete it

File: gluetool/glue.py (hunk 2 above)
Comment (happz) [reply gluetool/glue.py:390]: True, but this particular property definitely cannot be `None`, it is expected to be redefined by each and every module. Using `None` in this place is for the sake of initialization of the property.

File: gluetool/glue.py (hunk 3 above)
Comment (happz) [reply gluetool/glue.py:405]: In this context, it means "any number and type of arguments". Describing `options` and its parts in terms of types is not trivial, therefore many of methods working with options accept these "placeholder" types. I intend to make them more strict in the future, of course.

File: gluetool/glue.py (hunk 4 above)
Comment (happz) [reply gluetool/glue.py:701]: Right, comment added in the following patch.

File: gluetool/glue.py (hunk 5 above)
Comment (happz) [reply gluetool/glue.py:808]: The problem is that the type of `DryRunLevels.DEFAULT` (and others) is changed dynamically - they are defined as integers, and after that `enum` class grabs them and changes them to other objects with properties like `name`. I guess I could use `cast` here, to tell `mypy` that `self.dryrun_level` is not a mere integer but some type from `enum`  package, but I didn't try yet. There may be other options as well, for the meantime I opted for `ignore` - all `ignore` tags **will** be reviewed later :)

File: gluetool/log.py (hunk 6 above)
Comment (happz) [reply gluetool/log.py]: Right, fixed in the next patch.

File: gluetool/glue.py (hunk 2 above)
Comment (mkluson) [reply gluetool/glue.py:390]: ok

File: gluetool/__init__.pyi (hunk 1 above)
Comment (happz) [reply gluetool/__init__.pyi]: It is a leftover from one of my previous attempts, it is not necessary anymore. I removed it.

File: gluetool/glue.py (hunk 3 above)
Comment (happz) [reply gluetool/glue.py:405]: @mkluson are you satisfied with the explanation? Can we resolve the conversation?

Comment (mkluson) [reply gluetool/glue.py:405]: yes, thanks

**Review by mkluson — DISMISSED** [review]

**Comment by thrix** [comment]
superega giga change, I wonder how will python3 support influence this. 

I am still thinking if we should not go for Python3 support only later

**Comment by happz** [comment]
Well, this patch is in the direction of Python 3 anyway - it is not necessary to have type annotations to support Python 3, it is fully optional for both Python 2 and 3, but I'd like to add it anyway since with Python 3 features this would be one of the most tempting. And it'd help a lot even without Python 3 with the static analysis.

**Pull request merged**